README.md

```diff
@@ -393,6 +395,76 @@ BENCHMARK_REGISTER_F(MyFixture, BarTest)->Threads(2);
 /* BarTest is now registered */
 ```
 
+
+## User-defined counters
+
+You can add your own counters with user-defined names. The example below
+will add columns "Foo", "Bar" and "Baz" in its output:
+
+```c++
+static void BM_UserCountersExample(benchmark::State& state) {
+  double numFoos = 0, numBars = 0, numBazs = 0;
+  while (state.KeepRunning()) {
+    // ... count Foo,Bar,Baz events
+  }
+  state.SetCounter("Foo", numFoos);
+  state.SetCounter("Bar", numBars);
+  state.SetCounter("Baz", numBars);
+}
+```
+
+If you are compiling in C++11 mode or later, then overloads with
+std::initializer_list will be enabled. The calls to `SetCounter()` in the
+example above could then be written instead in a single call:
+
+```c++
+  state.SetCounter({{"Foo", numFoos}, {"Bar", numBars}, {"Baz", numBazs}});
+```
+
+`state.SetCounter()` works
+as expected with new or previously existing counters: if the counter already
+exists, then its value will be overwritten. If the counter does not exist,
+then `state.SetCounter()` will create a counter initialized with the provided
+value.
+
+You can mark counters as rates and/or as per-thread averages:
+
+```c++
+  // Set the counter as a rate. It will be presented divided
+  // by the duration of the benchmark.
+  state.SetCounter("FooRate", numFoos, benchmark::Counter::kIsRate);
+
+  // Set the counter as a thread-average quantity. It will
+  // be presented divided by the number of threads. */
+  state.SetCounter("FooRate", numFoos, benchmark::Counter::kAvgThreads);
+
+  // the third argument is a bitwise mask, so you can combine.
+  // This will mark as a thread-average rate:
+  state.SetCounter("FooRate", numFoos, benchmark::Counter::kIsRate|benchmark::Counter::kAvgThreads);
+```
+
+Note that setting the counter flags only works on the first call to `state.SetCounter()`; any subsequent values will be silently ignored.
+
+The return value of `state.SetCounter()` is of type `size_t` and is the handle
+to the counter. You can use this handle to get the counter with the method
+`state.GetCounter()`. You can also get the counter by name, either with
+`const char*` or `std::string const&`; this will do a string lookup, linear
+in the number of counters. The counter object has compound
+assignment overloads `+=`, `-=`, `*=`, `/=`. You can then change the counter
+as needed. For example:
+
+```c++
+static void BM_UserCountersExample(benchmark::State& state) {
+  state.SetCounter("LighterCounter"); // counter value defaults to 0
+  size_t handle = state.SetCounter("HeavyCounter"); // also 0
+  while (state.KeepRunning()) {
+    state.GetCounter("LighterCounter") += numLighterEvents; // will do a string lookup, linear in number of counters
```

Comment:
GetCounter is going to incur overhead which, if it's in the context of a benchmark keeprunning loop is not a great idea.

the examples here should be best in class, ie:

```
while (state.KeepRunning()) {
  numLighterEvents++;
  bigNumEvents++;
}
state.GetCounter("LighterCounter") += numLighterEvents;
state.GetCounter(handle) += bigNumEvents;
```